### PR TITLE
Add network and storage adapter interfaces

### PR DESCRIPTION
**Pull request by acurrieclark**
#265 points out that using interfaces for our adapters, rather than forcing to extend our own abstract adapters, would make the creation of new adapters more accessible.

This adds an interface for both network and storage adapters and updates the `RepoConfig` type to accept them rather than the abstract type.

I don't think this is breaking, but would appreciate a review. Specifically, I don't know where it is best to leave the JSDoc comments; on the abstract class or the interface? Left them on both for now.

### CODE REVIEW

**Comment by alexjg** [comment]
I like this change. The abstract classes don't have any implementations and I don't think we rely on instanceof checks anywhere, should we just mark them as deprecated?

**Comment by acurrieclark** [comment]
That would be my instinct, yes.

I have done some tests, and it looks like the comments don't propagate from the interface to the concrete implementation if they are removed from the abstract class, so we might need to leave them on both.

**Comment by alexjg** [comment]
In that case let's leave comments on both but add a deprecation notice to the abstract classes using `@deprecated` with a `{@link}` to the interface to use instead.

**Comment by acurrieclark** [comment]
Links added with `@deprecated`. Looks how I would expect, but I am no expert at JSDoc.

File: packages/automerge-repo/src/network/NetworkAdapter.ts [hunk 1]
```diff
@@ -1,29 +1,24 @@
...
- *
- */
-export interface PeerMetadata {
-  storageId?: StorageId
-  isEphemeral?: boolean
-}
+import { NetworkAdapterInterface } from "./NetworkAdapterInterface.js"
 
 /** An interface representing some way to connect to other peers
+ * @deprecated
```
Comment (alexjg) [inline packages/automerge-repo/src/network/NetworkAdapter.ts:10]: I think this should be more like `@deprecated use {@link NetworkAdapterInterface} instead`. You can run `npx typedoc` to see what the output looks like.

**Comment by acurrieclark** [comment]
There are a few warnings from running `typedoc`, but none relating to this PR following your suggestions

**Pull request merged**